[core][adag] Separate the outputs of execute and execute_async to multiple refs or futures to allow clients to retrieve them one at a time (#46908)

jeffreyjeffreywang opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Currently, if `MultiOutputNode` is used to wrap a DAG's output, you get back a single `CompiledDAGRef` or `CompiledDAGFuture`, depending on whether `execute` or `execute_async` is invoked, that points to a list of all of the outputs. To retrieve one of the outputs, you have to get and deserialize all of them at the same time.

This PR separates the output of `execute` and `execute_async` to a list of `CompiledDAGRef` or `CompiledDAGFuture` when the output is wrapped by `MultiOutputNode`. This is particularly useful for vLLM tensor parallelism. Since all shards return the same results, we only need to fetch result from one of the workers.

## Related issue number

<!-- For example: "Closes #1234" -->
Resolves #46908 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jeffreyjeffreywang commented:
Hi Ray developers, I would love to have some help validating `test_execution_schedule_gpu.py` tests as they require a GPU device which I don't have one.

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1564,42 +1570,47 @@ def _execute_until(
+                    if self._max_execution_index not in self._result_buffer:
+                        self._result_buffer[self._max_execution_index] = {}
+                    self._result_buffer[self._max_execution_index][chan] = fetcher.read(
+                        timeout
```
`timeout` needs to be updated in this loop?

jeffreyjeffreywang replied:
Good catch, yeah, reading from each channel should deduct from the global timeout.

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1564,42 +1570,47 @@ def _execute_until(
-                timeout = max(timeout, 0)
+                start_time = time.monotonic()
+
+                for chan, fetcher in self._dag_output_fetchers.items():
```
Do we need to fetch from all channels? Can you please add documentation?

jeffreyjeffreywang replied:
Yes, this aligns with the original behavior. The original code has a single fetcher which encapsulates all output channels, and invoking `read` on that fetcher (`_dag_output_fetcher`) fetches results from all output channels. Now, we have a fetcher for each channel (`_dag_output_fetchers`), and each fetcher is only responsible for getting results for its channel. The reason for this refactor is that we will now be able to get result from a single output channel so that we are not forced to get all results from all channels. I'll add documentation to make it clearer.

ruisearch42 reviewed (commented):
Fist pass. Looks pretty nice. Thanks for the contribution!

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -630,7 +630,7 @@ def __init__(
         # All results with higher indexes have not been generated yet.
         self._max_execution_index: int = -1
-        self._result_buffer: Dict[int, Any] = {}
+        self._result_buffer: Dict[int, Dict[ChannelInterface, Any]] = {}
```
```suggestion
        # execution_index -> {channel -> result}
        self._result_buffers: Dict[int, Dict[ChannelInterface, Any]] = {}
```

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1545,6 +1550,7 @@ def run(self):
     def _execute_until(
         self,
         execution_index: int,
+        channel: ChannelInterface,
```
can you update a docstring?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1564,42 +1570,48 @@ def _execute_until(
         TODO(rui): catch the case that user holds onto the CompiledDAGRefs
         """
-        from ray.dag import DAGContext
+        if self._max_execution_index < execution_index:
```
why do we need this if now?

jeffreyjeffreywang replied:
An ImportError will occur with the original code when destructing unused CompiledDAGRef (ones that we have not called get()). Here is a minimal repro for the original behavior:

```
foo = Foo.remote()
bar = Bar.remote()

with InputNode() as inp:
    dag = MultiOutputNode([foo.increment.bind(inp), bar.decrement.bind(inp)])

dag = dag.experimental_compile()

ref1 = dag.execute(1)
ref2 = dag.execute(1)

assert ref2.get() == [2, -2]

dag.teardown()
# When exiting the program, ref1.__del__ is invoked.
# Since it has not been called with get(), ref1.get() will
# be invoked subsequently. In _execute_until, even though
# the DAG won't be executed again, we still attempt to
# import a library, namely ray.dag.DAGContext. Since
# Python is shutting down, an ImportError occurs
# (ImportError: sys.meta_path is None, Python is likely
# shutting down).
```

The same problem persists with the current behavior if the if clause is not introduced. Another the benefit of introducing the if clause is to only import the library and calculate `timeout` when necessary. `timeout` is only used when `self._max_execution_index < execution_index` is `True`. We shouldn't import the library when `self._max_execution_index == execution_index`. I should keep the following while loop out of the if clause though.

jeffreyjeffreywang replied:
As I speak, I found out that if the last `CompiledDAGRef` is unused, the `ImportError` persists even when the if clause is introduced. Is there a guidance for Ray users to always invoke `get()` on all `CompiledDAGRef`? If not, this may be a bug that needs to be addressed in a separate PR. WDYT?

Here is a minimal repro:
```
foo = Foo.remote()
bar = Bar.remote()

with InputNode() as inp:
    dag = MultiOutputNode([foo.increment.bind(inp), bar.decrement.bind(inp)])

dag = dag.experimental_compile()

ref1 = dag.execute(1)
ref2 = dag.execute(1)

assert ref1.get() == [1, -1]

dag.teardown()
# Upon destruction, the DAG will be executed until the latest index. Again,
# we attempt to import DAGContext during program exit. 
```

jeffreyjeffreywang replied:
For now, I think it's worth keeping the if clause and have another bug tracking the dangling `CompiledDAGRef` issue.

jeffreyjeffreywang replied:
I think this issue I mentioned is correlated with your second point. Is there any reason why we want to avoid execution result leak (which is why we currently execute the DAG until the latest index and `get()` all results)?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1564,42 +1570,48 @@ def _execute_until(
-            if self._max_execution_index + 1 == execution_index:
-                # Directly fetch and return without buffering
+            while self._max_execution_index < execution_index:
+                if len(self._result_buffer) >= self._max_buffered_results:
```
I think this is wrong now. We should do len(self._result_buffer) * num_output_channels

rkooo567 replied:
can you also add a unit test? (or modify existing test)

jeffreyjeffreywang replied:
Agree. `len(self._result_buffer)` indicates the number of DAG executions while `len(self._result_buffer) * num_output_channels` represents the total number of outputs. I will adjust the naming of `_max_buffered_results` to `_max_buffered_executions` (any other suggestions?) and add a unit test.

jeffreyjeffreywang replied:
Actually, I think we need to change `max_buffered_results` to `max_buffered_executions` if we were to change `_max_buffered_results` to `_max_buffered_executions`. Since there are quite a few references to `max_buffered_results` in the repo (including developer APIs and `python/ray/dag/context.py` which doesn't have much relevance to the changes in this PR), I'd suggest to not modify the naming now. We can track it as a bug and address it in a follow-up.

As for the unit test, I'll add one tomorrow.

jeffreyjeffreywang replied:
Done adding a unit test for cases when there are multiple outputs. Now, if not all results from an execution index are fetched with `get()`, that execution index will still count towards the number of buffered results. Let me know if this sounds okay to you!

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1564,42 +1570,48 @@ def _execute_until(
+                # Fetch results from each output channel up to execution_index and store
+                # them separately to enable individual retrieval
+                for chan, fetcher in self._dag_output_fetchers.items():
+                    if self._max_execution_index not in self._result_buffer:
```
can you just use

```
self._result_buffer = defaultdict(dict)
```

and we can remove this code

jeffreyjeffreywang replied:
Sure thing, will do!

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1564,42 +1570,48 @@ def _execute_until(
+                for chan, fetcher in self._dag_output_fetchers.items():
+                    if self._max_execution_index not in self._result_buffer:
+                        self._result_buffer[self._max_execution_index] = {}
+                    self._result_buffer[self._max_execution_index][chan] = fetcher.read(
```
QQ @ruisearch42 do you know if there is any regression by doing this? (now we are blocking read each channel as many as output ref times)

jeffreyjeffreywang replied:
No longer applicable as `self._dag_output_fetchers` is gotten rid off entirely. There is no longer a fetcher for each channel. Results from all channels for each execution index are fetched at the same time and cached in `_result_buffer`.

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1690,13 +1706,20 @@ async def execute_async(
+            channel_futs = {}
+            for chan in self._dag_output_fetchers:
+                fut = asyncio.Future()
+                await self._fut_queues[chan].put(fut)
```
QQ: can we use asyncio.gather instead and do all put together?

jeffreyjeffreywang replied:
No longer applicable as `dag_output_fetchers` is gotten rid of. Per @stephanie-wang's comment, for each execution index, I fetch results from all channels and cache them in `_result_buffer` rather than fetching results separately for each channel with their individual fetcher (which was what `_dag_output_fetchers` represent).

rkooo567 reviewed (commented):
I think generally the approach makes sense! But I am seeing 2 problems.

1. It is backward incompatible changes and will break vllm. How complicated is it to support `MultiOutputNode(_multiple_return_refs=True)`? I am not suggesting to do it, but I'd like to measure the complexity 
2. I think although we don't call ray.get, ray.get is called and deserialization still happens when the dag ref is deallocated because of this code https://github.com/ray-project/ray/blob/4aea49f5301ac083aed5d016919482afd2f54f83/python/ray/experimental/compiled_dag_ref.py#L80. I think we should improve this part of code to not doing deserialization. I think this part is a little bit tricky to handle (so we can probably handle as a follow up)

jeffreyjeffreywang commented:
Hey @rkooo567, thank you so much for the review. Let me experiment a few things you suggested and get back to you tomorrow.

stephanie-wang reviewed (commented):
Thanks for this contribution!

jeffreyjeffreywang commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -226,10 +226,10 @@ def test_actor_methods_execution_order(ray_start_regular):
     # test that double_and_inc() is called after inc() on actor1
-    assert result == [4, 1]
+    assert ray.get(refs[0]) == 4
+    assert ray.get(refs[1]) == 1
```
It seems like `ray.get` only supports `CompiledDAGRef`, `ObjectRef`, and a list of `ObjectRef`, and `CompiledDAGRef` is not a subclass of `ObejctRef` although it's declared to be so in its docstring. I'm not sure how `ray.get(refs)` will work. Am I missing anything?

jeffreyjeffreywang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1633,9 +1645,13 @@ def execute(
         self._execution_index += 1
-        return ref
+
+        return refs[0] if len(refs) == 1 else refs
```
Great point, I'll adjust the logic.

jeffreyjeffreywang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -563,7 +563,7 @@ def __init__(
 
         self._enable_asyncio: bool = enable_asyncio
-        self._fut_queue = asyncio.Queue()
+        self._fut_queues: Dict[ChannelInterface, asyncio.queues.queue] = {}
```
Note: `self._result_buffer` and perhaps `self._dag_output_fetchers` need refactoring.

jeffreyjeffreywang replied:
`self._fut_queues` and `self._dag_output_fetchers` are changed to lists with indexing consistent with `self.dag_output_channels`. `self._result_buffer` is still a dict but no longer uses `ChannelInterface` for indexing (instead, `{execution_index->{channel_index->result}}`). The reason I kept it as a dict is that when we get and remove results from the buffer, the remaining results won't be out of order.

An alternative to use list for `self._result_buffer` is to mark the popped results as `None` instead of remove them from the buffer. However, the downside is that the size of the list remains the same even if we get all results from the buffer. From debugging perspective, I don't see a clear advantage for one over the other. Let me know if you prefer `self._result_buffer` to be a list as well.

jeffreyjeffreywang commented:
Thanks for all of the feedback. I'll keep addressing comments tomorrow and over the weekend. Will ping you as soon as the revision is ready for review.

jeffreyjeffreywang commented:
> It is backward incompatible changes and will break vllm. How complicated is it to support MultiOutputNode(_multiple_return_refs=True)? I am not suggesting to do it, but I'd like to measure the complexity.

@rkooo567, I think this should be quite straightforward to support. Essentially, we can ask clients to specify `multiple_return_refs=True` when they want a list of refs rather than a single ref wrapping all outputs. `execute` and `execute_async` can check whether `MultiOutputNode` is used to wrap the output and whether `multiple_return_refs` is set to True. If both conditions are satisfied, we return a list of refs. Otherwise, fall back to original behavior. I'll get the change in for you to review.

jeffreyjeffreywang commented:
Please hold off reviewing as I'm now working on making it backward compatible. Thanks!

jeffreyjeffreywang commented:
The change is now backward compatible. Working on unit tests and addressing the remaining comments.

jeffreyjeffreywang commented:
All comments are addressed except for a test for validating an error is raised when the number of buffered results exceeds `_max_buffered_results` which I will complete tomorrow. @rkooo567 @stephanie-wang The rest of the PR is ready for another round of review. Thank you 🙂

jeffreyjeffreywang commented:
Done with adding a new unit test for validating max buffered result logic. @stephanie-wang @rkooo567 please let me know if you have any other concerns with the latest changes. Thanks!

stephanie-wang commented:
> Done with adding a new unit test for validating max buffered result logic. @stephanie-wang @rkooo567 please let me know if you have any other concerns with the latest changes. Thanks!

Sorry for the delay, taking a look today!

rkooo567 commented:
yep will finish review by today as well!!

stephanie-wang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1864,11 +1909,62 @@ def _check_inputs(self, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -> None:
+        self,
+        execution_index: int,
+        channel_index: Optional[int],
+        result: Optional[Any] = None,
```
I would split this into a cache setter and getter method to make this logic more robust, instead of allowing `result` to be optional.

Ideally, `_execute_until` should also use these methods.

stephanie-wang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1864,11 +1909,62 @@ def _check_inputs(self, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -> None:
+                self._result_buffer[execution_index][chan_idx] = res
+
+        if channel_index is None:
+            del self._result_buffer[execution_index]
```
Doesn't this need to set `result`? Please add a test if not tested already.

jeffreyjeffreywang replied:
Yes, this should. This could potentially return incorrect results when CompiledDAGFutures are fetched in order. Will add a test as well.

stephanie-wang commented on python/ray/experimental/compiled_dag_ref.py:
```diff
@@ -148,9 +166,17 @@ def __await__(self):
         self._fut = None
 
-        return_vals = yield from fut.__await__()
+        if self._dag._has_execution_results(self._execution_index):
```
Nice change, this looks pretty clean!

One nit: for readability, change the logic to:

```
if not cached:
  yield from fut
  cache values
pop from cache
```

stephanie-wang reviewed (commented):
Overall structure looks good! Just some comments to unify the sync/async codepaths more, and ideally can we remove `self._multiple_return_refs`?

jeffreyjeffreywang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -696,6 +701,10 @@ def _get_creator_or_proxy_actor() -> "ray.actor.ActorHandle":
         return self._has_single_output
 
+    @property
+    def multiple_return_refs(self):
```
Will do!

jeffreyjeffreywang replied:
No longer applicable as this attribute is removed from CompiledDAG. I'm now only using `_has_single_output` as the flag to determine whether we want to return a list of refs/futures. Adjusted the property `multiple_return_refs` to `return_multiple_refs` for `MultiOutputNode` though.

jeffreyjeffreywang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1753,61 +1765,84 @@ def run(self):
-        # CompiledDAGRef guarantees that the same execution index will not
-        # be requested multiple times
-        return self._result_buffer.pop(execution_index)
+        # CompiledDAGRef guarantees that the same execution index and
```
Yup no problem. I agree with being overly cautious although this path typically doesn't induce errors.

jeffreyjeffreywang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1841,7 +1876,17 @@ def execute(
+        # MultiOutputNode is used to wrap outputs and client specified that a list of
+        # CompiledDAGRef, with each item corresponding to an output fetched from each
+        # output channel, should be returned.
+        if not self._has_single_output and self._multiple_return_refs:
```
Yep, removed `multiple_return_refs`.

stephanie-wang replied:
Hmm sorry I didn't quite understand this. I think the two cases are:

DAG contains a MultiOutputNode -> `_has_single_output=True` (and `returns_multiple_refs=True`)
DAG does not contain a MultiOutputNode -> `_has_single_output=False` (and `returns_multiple_refs=False`)

Can you explain which case is the one where `_has_single_output != returns_multiple_refs`?

jeffreyjeffreywang replied:
- DAG contains a MultiOutputNode `node`
    - `node.returns_multiple_refs=True`: `has_single_output=True`, `returns_multiple_refs=True`
    - `node.returns_multiple_refs=False`: `has_single_output=False`, `returns_multiple_refs=False`
- DAG does not contain a MultiOutputNode
    - `has_single_output=True`, `returns_multiple_refs=False`

Here is a minimal repro:
```
a = Actor.remote(0)
with InputNode() as i:
    dag = a.echo.bind(i)

compiled_dag = dag.experimental_compile()
print(compiled_dag._has_single_output) # True
print(compiled_dag._returns_multiple_refs) # False
```

Let me know if this makes sense to you 🙂

stephanie-wang replied:
I would simplify it to these cases:
- DAG contains a MultiOutputNode (which may only have one ref): `has_single_output=False`
- DAG does not contain a MultiOutputNode: `has_single_output=True`

Basically the first case should always return a list of refs, even if there is only one ref. You could consider renaming it to something like `returns_list`, to make it clearer.

jeffreyjeffreywang replied:
I introduced `returns_multiple_refs` as an attribute for MultiOutputNode per my discussion with SangBin.

> It is backward incompatible changes and will break vllm. How complicated is it to support MultiOutputNode(_multiple_return_refs=True)? I am not suggesting to do it, but I'd like to measure the complexity.

`execute` or `execute_async` now return a single ref/future even if there is a MultiOutputNode as long as the node's `returns_multiple_refs` attribute is set to False (which is by default). This prevents from breaking existing consumer code as they expect to get a single ref/future from `execute` and `execute_async`. Only new consumer code which specifies `returns_multiple_refs=True` will get a list of ref/future when there is a MultiOutputNode. Do you think this makes sense?

stephanie-wang replied:
Thanks for the explanation! I synced with @rkooo567 offline and we agreed that it should be okay to go ahead with this change and remove `_multiple_return_refs` (aDAGs are still in alpha so for now breaking changes are okay). @rkooo567 can shepherd the vLLM change required.

rkooo567 replied:
sorry @jeffreyjeffreywang you have to be back and fourth here :(.

jeffreyjeffreywang replied:
@rkooo567 No worries! I haven't taken the fact that aDAG being alpha into account either. @stephanie-wang thank you for the update, I'll revert the changes related to `_multiple_return_refs` later today.

jeffreyjeffreywang replied:
Done reverting :)

jeffreyjeffreywang commented:
> Overall structure looks good! Just some comments to unify the sync/async codepaths more, and ideally can we remove self._multiple_return_refs?

Thanks for the feedback, Stephanie! I refactored the code, unified sync/async codepaths, and added one more test case. I think we still need `multiple_return_refs` though (see my comments in the relevant thread).

Here are some pending issues that we might want to address as follow-ups according to our discussion above:

- [ ] Deserialization still happens in the destructor of CompiledDAGRef/CompiledDAGFuture
- [ ] ray.get doesn't support fetching results from a list of CompiledDAGRef
- [ ] There isn't sufficient ADAG test coverage for async cases (quite comprehensive for sync cases though)

stephanie-wang commented:
This looks great!

Two last things before we merge:
- could you answer my question in the other thread about why we need returns_multiple_refs?
- "ray.get doesn't support fetching results from a list of CompiledDAGRef" -> can we support this? Otherwise we will have a breaking API change. A naive solution that just `gets` the refs in order is good for now.

jeffreyjeffreywang commented:
> "ray.get doesn't support fetching results from a list of CompiledDAGRef" -> can we support this? Otherwise we will have a breaking API change. A naive solution that just gets the refs in order is good for now.
Definitely, will do 😃

jeffreyjeffreywang commented:
Done supporting fetching a list of CompiledDAGRef with `ray.get()`. @stephanie-wang, let me know if you have further concerns. Thank you again for all of your feedback!

stephanie-wang commented on python/ray/_private/worker.py:
```diff
@@ -2650,6 +2662,15 @@ def get(
             return object_refs.get(timeout=timeout)
 
+        if isinstance(object_refs, list):
+            allCompiledDAGRefs = True
```
```suggestion
            all_compiled_dag_refs = True
```
(we use underscore naming for python)

stephanie-wang commented on python/ray/_private/worker.py:
```diff
@@ -2650,6 +2662,15 @@ def get(
+                allCompiledDAGRefs = allCompiledDAGRefs and isinstance(
+                    object_ref, CompiledDAGRef
+                )
+            if allCompiledDAGRefs:
+                return [object_ref.get(timeout=timeout) for object_ref in object_refs]
```
I think we also need to handle the else condition. It's okay to throw an error for now if the list contains a CompiledDAGRef but not all CompiledDAGRefs.

jeffreyjeffreywang commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1447,11 +1447,6 @@ def _get_or_compile(
         assert [
             output_channel is not None for output_channel in self.dag_output_channels
         ]
-        # If no MultiOutputNode was specified during the DAG creation, there is only
```
Revert

stephanie-wang reviewed (approved):
Awesome, thank you for this great contribution!

By the way, if you are not already on it, please feel free to join #ray-accelerated-dags at ray.slack.com. I'm sure there are some interested users there :)

jeffreyjeffreywang commented:
Thank you so much @stephanie-wang for all of the reviews. I'm looking forward to contributing more 😄 

Would you mind taking a final look at my latest update? I resolved some merge conflicts with SangBin's recent changes.

jeffreyjeffreywang commented:
> I think although we don't call ray.get, ray.get is called and deserialization still happens when the dag ref is deallocated because of this code. https://github.com/ray-project/ray/blob/4aea49f5301ac083aed5d016919482afd2f54f83/python/ray/experimental/compiled_dag_ref.py#L80
> I think we should improve this part of code to not doing deserialization. I think this part is a little bit tricky to handle (so we can probably handle as a follow up)

Tracking this issue with #47614.

stephanie-wang commented:
> Thank you so much @stephanie-wang for all of the reviews. I'm looking forward to contributing more. 😄
> 
> Would you mind taking a final look at my latest update? I resolved some merge conflicts with SangBin's recent changes.

It looks good to me! Will merge once tests pass.

rkooo567 commented:
also to be clear @stephanie-wang it will be a breaking change right?

jeffreyjeffreywang commented:
> also to be clear @stephanie-wang. @jeffreyjeffreywang it will be a breaking change right?

Correct, this will be breaking. `execute` and `execute_async` on adag with `MultiOutputNode` will always return a list of refs/futures.

rkooo567 commented:
makes sense! Also @jeffreyjeffreywang do we have issues for all follow up problems?

```
- Deserialization still happens in the destructor of CompiledDAGRef/CompiledDAGFuture
-  ray.get doesn't support fetching results from a list of CompiledDAGRef
-  There isn't sufficient ADAG test coverage for async cases (quite comprehensive for sync cases though)
```

jeffreyjeffreywang commented:
> makes sense! Also @jeffreyjeffreywang do we have issues for all follow up problems?

Yup, the second one is addressed in this PR. #47662 and #47614 track the remaining follow-up problems.

rkooo567 commented:
@jeffreyjeffreywang https://github.com/ray-project/ray/issues/47684

actuallay is this a known issue? can you take a look?

jeffreyjeffreywang commented:
@rkooo567 Yeah, I also realized this may be a problem over the weekend. I think we need to support calling `ray.wait` on `CompiledDAGFuture` and a list of `CompiledDAGFuture` as well.
